Kafka Adapter Performance Enhancement

zliu41 opened the pull request:
- Now using the avg time to pull a record for mapper load balancing, instead of avg record size. This is more accurate since the time spent on a record is not necessarily proportional to its size. This is done in the following way:
 - Each partition pulled in the previous run has an avg time per record in the workunitstate.
 - In the next run, the estimated avg time per record for each topic is the geometric mean of the avg time per record of all partitions. For example if a topic has two partitions whose avg time per record in the previous run are 2 and 8, the next run will use 4 as the estimated avg time per record. The reason to choose **geometric mean** over algebraic mean is because large numbers are likely outliers, e.g., a topic may have 5 partitions, and the avg time per record collected from the previous run could sometimes be [1.1, 1.2, 1.1, 1.3, 100].
 - If a topic was not pulled in the previous run, its estimated avg time per record is the geometric mean of the estimated avg time per record of all topics that were pulled in the previous run. If no topic was pulled in the previous run, then we use a default value 1.0.

- Previously `KafkaSource` uses a bi-level bin packing approach for assigning workunits to mappers. Now added an option to use single-level bin packing, which is the default option. Bi-level bin packing has two advantages: (1) reduce the number of small output files since it tends to pack partitions of the same topic together; (2) reduce the total number of workunits / tasks since multiple partitions of the same topic are assigned to the same task. A task has a non-trivial cost of initialization, tear down and task state persistence. **However**, bi-level bin packing has more mapper skew than single-level bin packing, because if we pack lots of partitions of the same topic to the same mapper, and we underestimate the avg time per record for this topic, then this mapper could be much slower than other mappers. In my observation, the advantages of single-level bin packing outweigh the disadvantages.
 - Note that for each skipped topic, we create an empty workunit for each partition to preserve its checkpoints. In single-level bin packing mode, we still assign all partitions of a skipped topic to the same workunit / task, so that a single empty task will be created for this topic, instead of one empty task per partition. So single-level bin packing is more like 1.5-level bin packing.

- Modified the way mapper load is calculated: added a new factor of `log(number of tasks assigned to this mapper)`. This is because, as said above, each task is associated with a non-trivial cost. So, running 1000 small tasks in a mapper takes much longer than running 3 big tasks with the same amount of data.

- Previously the existence of topic schemas was verified in the mapper, so if a topic has no schema, we would still create workunits for it as usual, e.g., create 100 huge workunits which occupy 100 mappers, and these mappers ended up doing nothing. Now this is verified in the driver and disqualified topics will be skipped.

- Some other small changes and refactoring, including changing the "Queue stats of fork" log message to debug level since it sometimes floods the mapper log.

Tested on Nertz and this has lead to a significant reduction in the time each run takes.

zliu41 commented:
@chavdar could you review?

chavdar commented on gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java:
```diff
@@ -154,76 +164,120 @@ private void createEmptyWorkUnitsForSkippedPartitions(Map<String, List<WorkUnit>
       }
     }
-    long avgGroupSize = (long) ((double) totalEstDataSize / (double) numOfMultiWorkunits / 3.0);
+    if (isWorkUnitPreGroupingEnabled(state)) {
```
Can you abstract-away the bin-packing strategy with an interface and a couple of implementing classes. This will make adding new strategies easier in the future and the code cleaner?

chavdar commented:
Is it possible to emit some metric to measure the effectiveness of the bin-packing algorithm? Say, comparing the prediction with the observed time after the run is done.

zliu41 commented:
@chavdar updated. Added the following interfaces and classes: 
- `KafkaWorkUnitPacker`
 - `KafkaSingleLevelWorkUnitPacker`
 - `KafkaBiLevelWorkUnitPacker`
- `KafkaWorkUnitSizeEstimator`
 - `KafkaAvgRecordTimeBasedWorkUnitSizeEstimator`
 - `KafkaAvgRecordSizeBasedWorkUnitSizeEstimator`

For metrics, we can't (and don't need to) estimate the exact time each mapper takes. We only estimate the relative load among different mappers, which hopefully is proportional to their running times. So I've added the min estimated load and max estimated load of all mappers to the state, so that MetricReport reporter can report it to MetricReport which can then be compared against the actual mapper running time.

chavdar commented:
Mostly minor comments. Overall, LGTM
